processing/src/main/java/io/druid/query/PrioritizedExecutorService.java

```diff
@@ -185,89 +190,215 @@ public int getQueueSize()
+  {
+    this.delegate = delegate;
+    this.priority = priority;
+  }
 
-    PrioritizedListenableFutureTask(ListenableFutureTask<V> delegate, int priority)
-    {
-      this.delegate = delegate;
-      this.priority = priority;
-    }
+  @Override
+  public void run()
+  {
+    delegate.run();
+  }
 
-    @Override
-    public void run()
-    {
-      delegate.run();
-    }
+  @Override
+  public boolean cancel(boolean mayInterruptIfRunning)
+  {
+    return delegate.cancel(mayInterruptIfRunning);
+  }
 
-    @Override
-    public boolean cancel(boolean mayInterruptIfRunning)
-    {
-      return delegate.cancel(mayInterruptIfRunning);
-    }
+  @Override
+  public boolean isCancelled()
+  {
+    return delegate.isCancelled();
+  }
 
-    @Override
-    public boolean isCancelled()
-    {
-      return delegate.isCancelled();
-    }
+  @Override
+  public boolean isDone()
+  {
+    return delegate.isDone();
+  }
 
-    @Override
-    public boolean isDone()
-    {
-      return delegate.isDone();
-    }
+  @Override
+  public V get() throws InterruptedException, ExecutionException
+  {
+    return delegate.get();
+  }
 
-    @Override
-    public V get() throws InterruptedException, ExecutionException
-    {
-      return delegate.get();
-    }
+  @Override
+  public V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException
+  {
+    return delegate.get(timeout, unit);
+  }
 
-    @Override
-    public V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException
-    {
-      return delegate.get(timeout, unit);
-    }
+  @Override
+  public void addListener(Runnable listener, Executor executor)
+  {
+    delegate.addListener(listener, executor);
+  }
 
-    @Override
-    public void addListener(Runnable listener, Executor executor)
-    {
-      delegate.addListener(listener, executor);
-    }
+  @Override
+  public int getPriority()
+  {
+    return priority;
+  }
+
+  @Override
+  public int compareTo(PrioritizedListenableFutureTask otherTask)
+  {
+    return Ints.compare(otherTask.getPriority(), getPriority());
+  }
+}
 
+class PrioritizedTaskBlockingQueue extends PriorityBlockingQueue<Runnable>
+{
+  private static final Comparator<Runnable> PRIORITY_AND_ENUMERATION_COMPARABLE = new Ordering<Runnable>()
+  {
     @Override
-    public int getPriority()
+    public int compare(@Nullable Runnable r1, @Nullable Runnable r2)
     {
-      return priority;
+      int compare = 0;
+      if (r1 instanceof Prioritized && r2 instanceof Prioritized) {
+        compare = Integer.compare(((Prioritized) r1).getPriority(), ((Prioritized) r2).getPriority());
+      }
+      if (compare == 0) {
+        if (r1 instanceof Enumerated && r2 instanceof Enumerated) {
+          compare = Long.compare(((Enumerated) r1).getEnumeration(), ((Enumerated) r2).getEnumeration());
+        }
+      }
+      return compare;
     }
+  }.reverse(); // Higher means more priority
```

Comment:
that would be a reverse. IMHO it's more disingenuous to swap the operators in compare than to explicitly call reverse, because the former can look like an unintended situation.